### PR TITLE
ATA pipeline dev: mode_b_config.h #define for static values of BLADE configuration

### PR DESCRIPTION
**Pull request by RocketRoss**
This replaces #7.

The added `mode_b_config.h` #defines values critical to the ATA pipeline configuration. This better enables an external script to reconfigure the BLADE pipeline (before recompiling of course).
In particular `BLADE_ATA_MODE_B_OUTPUT_ELEMENT_T` switches between CF16 and CF32, in the latter case the pipeline's outputCast is mitigated.

This work also expects copy2D (#10) so that the output can be chunked and padded ready for network-packetisation. Will be kept as a draft until after #10.

### CODE REVIEW

**Comment by RocketRoss** [comment]
@luigifcruz This is ready for merge. Thanks.

**Comment by luigifcruz** [comment]
@RocketRoss Ops forgot to submit my comment. 😓

File: src/pipelines/ata/mode_b.cc [hunk 1]
```diff
@@ -8,31 +8,69 @@ ModeB::ModeB(const Config& config) : config(config) {
...
+    // output is un-casted beamformer output (CF32)
+    BL_CHECK(this->copy2D(
+        output,
+        BLADE_ATA_MODE_B_OUTPUT_MEMCPY2D_DPITCH,// dpitch
+        beamformer->getOutput(),                // src
+        BLADE_ATA_MODE_B_OUTPUT_MEMCPY2D_WIDTH, // spitch
+        BLADE_ATA_MODE_B_OUTPUT_MEMCPY2D_WIDTH, // width
+        (beamformer->getOutputSize()*sizeof(BLADE_ATA_MODE_B_OUTPUT_ELEMENT_T))/BLADE_ATA_MODE_B_OUTPUT_MEMCPY2D_WIDTH
+        ));
+    #endif
```
[the remark this answers is not in the document]
Comment (RocketRoss) [reply src/pipelines/ata/mode_b.cc:73]: Very close indeed. Only difference is the src pointer as the outputCast is not instantiated if the output_element_type is CF32..
I could rewrite as
```
    BL_CHECK(this->copy2D(
        output,
        BLADE_ATA_MODE_B_OUTPUT_MEMCPY2D_DPITCH,// dpitch
#if BLADE_ATA_MODE_B_OUTPUT_NCOMPLEX_BYTES != 8
        // source is cast output
        outputCast->getOutput(),                // src
#else
         // source is beamformer output (CF32)
        beamformer->getOutput(),                // src
#endif
        BLADE_ATA_MODE_B_OUTPUT_MEMCPY2D_WIDTH, // spitch
        BLADE_ATA_MODE_B_OUTPUT_MEMCPY2D_WIDTH, // width
        (beamformer->getOutputSize()*sizeof(BLADE_ATA_MODE_B_OUTPUT_ELEMENT_T))/BLADE_ATA_MODE_B_OUTPUT_MEMCPY2D_WIDTH
        ));
```

**Comment by luigifcruz** [comment]
Thanks for the PR! The logic seems great to me. My main concern are the hard coded definitions. Changing them would require a whole library recompilation. I think it would be better to move all configuration definitions to the `ModeB::Config` structure. This way, it would be possible to change parameters in runtime without recompilation. An exception to this would be the output type definition. In this case it can be defined as a template.

Let me know what you think.

I can implement these changes if you want or you can take a shot at it.

**Comment by luigifcruz** [comment]
I committed my suggestions to 1f7e54d8976a09fa87dae2a800893c48ae4bc309. Let me know what you think. I believe it delivers the same functionality with runtime personalization.

**Comment by RocketRoss** [comment]
Yeah, that looks far neater. Very nice! I'll pull and give it a test run today :+1:.

**Comment by RocketRoss** [comment]
It's perfect. Thanks Luigi!

One thing I have noticed is that in my external use case (Hpguppi_daq) I don't get any of the BL_INFO printouts asides from the banner (I've locally changed the BL_DEBUGs to BL_INFO). I see them when I run the internal C test though. Any clues?

**Comment by RocketRoss** [comment]
I guess you might want to merge from this PR...

**Comment by luigifcruz** [comment]
Awesome, thanks for testing!

I found a bug in my implementation. I committed the path in 48b4eca72d9ef0e90b62ff0c5d722b77df08f1ae. Can you cherry-pick this commit?

Regarding the logger messages, by default, it'll only print `BL_WARN` or worse messages if `NDEBUG` isn't defined. [This](https://github.com/luigifcruz/blade/blob/274453eb2495bb051767bb55dc9c990495d96fa3/include/blade/logger.hh#L8-L13) is the part of the code that does this. This `NDEBUG` define is automatically set when the code is compiled with debug options enabled. It might explain the behavior.

**Comment by luigifcruz** [comment]
Awesome, this looks good! Proceeding with the merge. Thanks, @RocketRoss!

**Pull request merged**